### ISSUE TITLE
Self Signed local registry

### ISSUE THREAD

**Issue by jrsmile** [vmware-samples/vcenter-event-broker-appliance#818]
Hi,

how can i add a local ca to the knative-servicing namespace. when deploying a funtion from a private registry i could login but the image is not fetched because veba does not know about our internal ca.

`{"severity":"INFO","timestamp":"2022-03-23T09:54:00.504813479Z","logger":"controller.event-broadcaster","caller":"record/event.go:282","message":"Event(v1.ObjectReference{Kind:\"Revision\", Namespace:\"vmware-functions\", Name:\"kn-ps-stackfield-00001\", UID:\"bd837104-677c-4c1d-8afb-0cdf049a34f5\", APIVersion:\"serving.knative.dev/v1\", ResourceVersion:\"1193523\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' Unable to fetch image \"gitlab.domain.de:5050/reiss/kn-ps-stackfield:latest\": failed to resolve image to digest: Get \"https://gitlab.domain.de:5050/v2/\": x509: certificate signed by unknown authority","commit":"7499ae2","knative.dev/pod":"controller-66cb4b556b-sfk7v"}`

i found some info about custom CAs but they are outdated and for the vcenter link and not for the registry.
[docs/kb/advanced-certificates.md](https://github.com/vmware-samples/vcenter-event-broker-appliance/blob/development/docs/kb/advanced-certificates.md)
[https://zhimin-wen.medium.com/running-knative-on-on-premise-kubernetes-cluster](https://zhimin-wen.medium.com/running-knative-on-on-premise-kubernetes-cluster-ibm-cloud-private-4227ee322aa0)

but i could not find the yaml for redeploying knative-servicing.

any easy to follow ideas?

Best Regards, J

**Comment by kremerpatrick** [comment]
I will check into this and get back to you.

**Comment by kremerpatrick** [comment]
Please try the following:

1) /etc/containerd/config.toml
- Create a backup of the file
- Find this section
 [plugins."io.containerd.grpc.v1.cri".registry]
      config_path = ""

Change the config_path to:
     config_path = "/etc/containerd/certs.d"

2) Create the certs.d directory
mkdir /etc/containerd/certs.d
chmod 755 /etc/containerd/certs.d

3) Create a directory for your registry. In this example, I am using a Harbor registry with a URL of harbor.set.local
mkdir /etc/containerd/certs.d/harbor.set.local
chmod 755 /etc/containerd/certs.d/harbor.set.local

4) Create the hosts.toml file
vi /etc/containerd/certs.d/harbor.set.local/hosts.toml
chmod 644 /etc/containerd/certs.d/harbor.set.local/hosts.toml

The file should look like this, make sure to use your registry's URL

[host."https://harbor.set.local"]
  capabilities = ["pull", "resolve", "push"]
  skip_verify = true

5) Restart containerd
systemctl restart containerd

6) Pull from your local registry - in this example, I dropped an nginx image into Harbor
crictl pull harbor.set.local/veba/nginx:latest

You should see success
Image is up to date for sha256:f2f70adc5d89aa922836e9cc6801980a12a7ff9012446cc6edf52ef8798a67bd

**Comment by kremerpatrick** [comment]
Related: the containerd documentation for this is not great - this post lead me down the right path
https://stackoverflow.com/questions/65681045/adding-insecure-registry-in-containerd

**Comment by kremerpatrick** [comment]
After doing this crictl pull works but I still can't deploy a function. Continuing to investigate - trying to use kn-ps-echo with the container hosted in the local registry, but still getting "Unable to get the Subscriber's URI" errors.

**Comment by kremerpatrick** [comment]
The documentation mentions the config change works with crictl, not ctr.

**Comment by kremerpatrick** [comment]
Anyway, William suggested that I do a kubectl -n vmware-functions get events and I still saw x.509 errors after making the above changes and getting crictl to work.  I then tried this to get Photon to trust the self-signed cert: https://github.com/vmware/photon/issues/592

The result of that is that the OS now trusts - I can do a curl https://harbor.set.local without errors. And the x.509 errors are gone from vmware-functions events. But I still can't get a function to deploy

**Comment by kremerpatrick** [comment]
I'm now playing with these posts http://hypernephelist.com/2021/03/23/kubernetes-containerd-certificate.html  https://paraspatidar.medium.com/add-self-signed-or-ca-root-certificate-in-kubernetes-pod-ca-root-certificate-store-cb7863cb3f87  to see if I can get it to work.

**Comment by jrsmile** [comment]
I will have a look too, thanks in advance.

**Comment by jrsmile** [comment]
noticed another henn egg problem: when trying to get c_rehash working  i used "tdnf provides */c_rehash" to get the packet in question. unfortunately we use a proxy for internet connection, the proxy is set corretly to intercept internet traffic and ignore traffic that goes diretly to the registry. but it reencrypts internet traffic with its own certificate also. 

`root@bn-vcveba-01 [ /etc/ssl/certs ]# tdnf provides */c_rehash
Refreshing metadata for: 'VMware Photon Linux 4.0 (x86_64) Updates'
retrying 1/10
retrying 2/10
retrying 3/10
retrying 4/10
retrying 5/10
retrying 6/10
retrying 7/10
retrying 8/10
retrying 9/10
retrying 10/10
Error(1261) : SSL peer certificate or SSH remote key was not OK
Error: Failed to synchronize cache for repo 'VMware Photon Linux 4.0 (x86_64) Updates' from 'https://packages.vmware.com/photon/4.0/photon_updates_4.0_x86_64'
Disabling Repo: 'VMware Photon Linux 4.0 (x86_64) Updates'
Refreshing metadata for: 'VMware Photon Linux 4.0 (x86_64)'
retrying 1/10
retrying 2/10
retrying 3/10
retrying 4/10
retrying 5/10
retrying 6/10
retrying 7/10
retrying 8/10
retrying 9/10
retrying 10/10
Error(1261) : SSL peer certificate or SSH remote key was not OK
Error: Failed to synchronize cache for repo 'VMware Photon Linux 4.0 (x86_64)' from 'https://packages.vmware.com/photon/4.0/photon_release_4.0_x86_64'
Disabling Repo: 'VMware Photon Linux 4.0 (x86_64)'
Refreshing metadata for: 'Kubernetes'
retrying 1/10
retrying 2/10
retrying 3/10
retrying 4/10
retrying 5/10
retrying 6/10
retrying 7/10
retrying 8/10
retrying 9/10
retrying 10/10
Error(1261) : SSL peer certificate or SSH remote key was not OK
Error: Failed to synchronize cache for repo 'Kubernetes' from 'https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64'
Error(1261) : SSL peer certificate or SSH remote key was not OK
`

so i have to add this cert as well.

as c_rehash is not available in the vm i hashed the pem's myself and added symlinks to /etc/ssh/certs with 
`openssl x509 -hash -fingerprint -noout -in gitlab.pem
d5f80f11
SHA1 Fingerprint=8E:D4:A9:36:A8:BD:E9:4E:8C:88:FA:25:1E:E2:A9:11:07:F9:89:82
`
`root@bn-vcveba-01 [ /etc/ssl/certs ]# ln -s gitlab.pem d5f80f11.0`

but no success.

**Comment by lamw** [comment]
@jrsmile @kremerpatrick I've just successfully configured my custom registry (Harbor) with VEBA. I'm currently writing up the instructions and will share that out shortly

tl;dr - You need to update BOTH containerd as well as the Knative Serving controller for this to work

**Comment by jrsmile** [comment]
These are great news, i'll try it out as soon as i can.

**Comment by lamw** [comment]
As promised, here's the detailed instructions for updating both containerd and Knative Serving controller to support custom registry https://williamlam.com/2022/03/how-to-configure-knative-and-containerd-in-vmware-event-broker-appliance-veba-to-use-a-custom-registry.html

If you've got credentials to your registry that you need to specify, I believe you just need to create an imagePullSecret in the `knative-serving` namespace as outlined https://knative.dev/docs/serving/deploying-from-private-registry/

**Comment by jrsmile** [comment]
hi, im currently trying your blogpost:

`sed -i "s/      \[plugins.\"io.containerd.grpc.v1.cri\".registry.mirrors\].*/      \[plugins.\"io.containerd.grpc.v1.cri\".registry.mirrors\]\n         \[plugins.\"io.containerd.grpc.v1.cri\".registry.configs.\"${REGISTRY}\".tls\]\n         ca_file = \"${CUSTOM_REGISTRY_CERT_PATH}\"/g" /etc/containerd/config.toml`

should be:

`sed -i "s/      \[plugins.\"io.containerd.grpc.v1.cri\".registry.mirrors\].*/      \[plugins.\"io.containerd.grpc.v1.cri\".registry.mirrors\]\n         \[plugins.\"io.containerd.grpc.v1.cri\".registry.configs.\"${CUSTOM_REGISTRY}\".tls\]\n         ca_file = \"${CUSTOM_REGISTRY_CERT_PATH}\"/g" /etc/containerd/config.toml`

im following along, if i find other problems ill document them here.

**Comment by jrsmile** [comment]
im done, was realy easy with your blogpost thank you very much,
anyway, im currently watching kubectl -n knative-serving get deployment/controller for changes but the controller has an age of 6days and nothing happens, do i have to force an update somehow?

**Comment by lamw** [comment]
> im done, was realy easy with your blogpost thank you very much, anyway, im currently watching kubectl -n knative-serving get deployment/controller for changes but the controller has an age of 6days and nothing happens, do i have to force an update somehow?

You could just delete the controller and it should force an update. After applying the edit, it should have terminated the container ...

**Comment by jrsmile** [comment]
it worked somehow
after issuing: `kubectl -n knative-serving delete deployment/controller`

nothing happened so i executed: `kubectl apply -f new-knative-serving-controller.yaml` again

the controller was recreated. 👍 

now i get the following message:

`{"severity":"ERROR","timestamp":"2022-03-29T18:53:02.428486569Z","logger":"controller","caller":"controller/controller.go:566","message":"Reconcile error","commit":"7499ae2","knative.dev/pod":"controller-77c577465-d7l2j","knative.dev/controller":"knative.dev.serving.pkg.reconciler.revision.Reconciler","knative.dev/kind":"serving.knative.dev.Revision","knative.dev/traceid":"a6f9f50d-9cfb-44b5-9fc7-f59aab2d9aec","knative.dev/key":"vmware-functions/kn-ps-stackfield-00001","duration":"87.148526ms","error":"failed to update deployment \"kn-ps-stackfield-00001-deployment\": Operation cannot be fulfilled on deployments.apps \"kn-ps-stackfield-00001-deployment\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"knative.dev/pkg/controller.(*Impl).handleErr\n\tknative.dev/pkg@v0.0.0-20211206113427-18589ac7627e/controller/controller.go:566\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20211206113427-18589ac7627e/controller/controller.go:543\nknative.dev/pkg/controller.(*Impl).RunContext.func3\n\tknative.dev/pkg@v0.0.0-20211206113427-18589ac7627e/controller/controller.go:478"}
{"severity":"INFO","timestamp":"2022-03-29T18:53:02.429354393Z","logger":"controller.event-broadcaster","caller":"record/event.go:282","message":"Event(v1.ObjectReference{Kind:\"Revision\", Namespace:\"vmware-functions\", Name:\"kn-ps-stackfield-00001\", UID:\"155bf72b-484f-41dd-b81a-833fd4532909\", APIVersion:\"serving.knative.dev/v1\", ResourceVersion:\"15076116\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' failed to update deployment \"kn-ps-stackfield-00001-deployment\": Operation cannot be fulfilled on deployments.apps \"kn-ps-stackfield-00001-deployment\": the object has been modified; please apply your changes to the latest version and try again","commit":"7499ae2","knative.dev/pod":"controller-77c577465-d7l2j"}
`

the status on the veba web-ui is:
 ConfigurationsReady : Unknown
Ready : Unknown
RoutesReady : Unknown 

maybe there is something left from previous tests?

**Comment by lamw** [comment]
I'd recommend deleting your function deployment (e.g. un-deploy your function) and once Knative Serving controller is ready, then you can deploy your function

**Comment by jrsmile** [comment]
i end up with:

Error: ImagePullBackOff

`Failed to pull image "gitlab.domain.de:5050/reiss/kn-ps-stackfield@sha256:d48e4f165620689a2c5bd041c46c4e5e120f60f6faaf03fae1c3c84be1df1f61": rpc error: code = Unknown desc = failed to pull and unpack image "gitlab.domain.de:5050/reiss/kn-ps-stackfield@sha256:d48e4f165620689a2c5bd041c46c4e5e120f60f6faaf03fae1c3c84be1df1f61": failed to resolve reference "gitlab.domain.de:5050/reiss/kn-ps-stackfield@sha256:d48e4f165620689a2c5bd041c46c4e5e120f60f6faaf03fae1c3c84be1df1f61": failed to do request: Head "https://gitlab.domain.de:5050/v2/reiss/kn-ps-stackfield/manifests/sha256:d48e4f165620689a2c5bd041c46c4e5e120f60f6faaf03fae1c3c84be1df1f61": Method Not Allowed`

i had configured the login credentails in the function.yaml like so:

`spec:
  template:
    metadata:
      annotations:
        autoscaling.knative.dev/maxScale: "1"
        autoscaling.knative.dev/minScale: "1"
    spec:
      containers:
        - image: gitlab.domain.de:5050/reiss/kn-ps-stackfield:latest
          envFrom:
            - secretRef:
                name: stackfield-secret
          env:
            - name: FUNCTION_DEBUG
              value: "false"
      imagePullSecrets:
      - name: regcred`

**Comment by lamw** [comment]
Great to hear @jrsmile 

I'll open up a separate issue to add this detail to our official documentation and reference blog post for automation purposes

**Comment by lamw** [comment]
Closing per last comment
